torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -0,0 +1,916 @@
+        from . import ConstantVariable, UserFunctionVariable
+        from .builder import wrap_fx_proxy
+
+        assert (
+            all(isinstance(value, ConstantVariable) for value in kwargs.values())
+            or not kwargs
+        ), "only constant kwargs are supported"
+
+        from . import AutogradFunctionVariable, TorchVariable
+
+        pre_side_effects = tx.output.side_effects.clone()
+        always_restore = self.value.__name__ == "trampoline_autograd_bwd"
+        if (
+            self.value.__name__ == "trampoline_autograd_bwd"
+            or self.value.__name__ == "trampoline_autograd_fwd"
+        ):
+            fn = UserFunctionVariable(self.value, source=self.source)
+        else:
+            fn = TorchVariable(self.value)
+        checkpoint = tx.copy_graphstate()
+        pre_guards = tx.output.guards
+        graph_checkpoint = tx.output.graph
+        (
+            body_r,
+            body_graph,
+            body_lifted_freevars,
+        ) = speculate_subgraph(
+            tx,
+            fn,
+            [
+                *args,
+            ],
+            graph_checkpoint,
+            checkpoint,
+            # Backwards should never, ever be stored!
+            always_restore=always_restore,
+        )
+        post_guards = tx.output.guards
+        if body_lifted_freevars:
+            for freevar in body_lifted_freevars.keys():
+                if "saved_tensor_marked" not in freevar.node.meta:
+                    unimplemented("NYI - freevars in autograd function.")
+
+        post_side_effects = tx.output.side_effects
+        if post_side_effects.diff(pre_side_effects):
+            diff = (
+                post_side_effects.id_to_variable.keys()
+                - pre_side_effects.id_to_variable.keys()
+            )
+            for d in diff:
+                if not isinstance(
+                    post_side_effects.id_to_variable[d].value,
+                    AutogradFunctionVariable,
+                ):
+                    unimplemented("NYI - side effects in autograd function.")
+
+        if always_restore:
+            if post_guards - pre_guards:
+                unimplemented("NYI - New guards discovered in a restoring state")
+            # Nothing left to do here
+            return None
+
+        p_args = (
+            *(arg.as_proxy() for arg in args),
+            *(arg for arg in body_lifted_freevars.keys()),
+        )
+        r = body_r.as_proxy().node.meta["example_value"]
+        example_value = r
+
+        _, p_kwargs = proxy_args_kwargs([], kwargs)
+
+        # Store the invocation as a call
+        return wrap_fx_proxy(
+            tx=tx,
+            proxy=tx.output.create_proxy(
+                "call_function",
+                self.value,
+                args=tuple(p_args),
+                kwargs=p_kwargs,
+            ),
+            example_value=example_value,
+        )
+
+
+class HigherOrderWrapVariable(TorchHigherOrderOperatorVariable):
```

Comment:
There's a lot of code duplication here:
- Everyone needs to check that "only constant kwargs are supported" -> this should probably be a helper function
- HigherOrderWrapVariable and HigherOrderCheckpointVariable do the same thing. HigherOrderCheckpointVariable and HigherOrderWrapVariable should probably inherit from a common base that has the call_function. There will be other HigherOrderOps in the future that behave similarly.

In general, we should try to reuse code among these implementations as much as possible so that future improvements (e.g. support for body_functions that accept pytrees) can easily be added across the higher order operators.